### DISCUSSION TITLE
LTG8F328x has 7 more digital I/O-s than ATmega328P

### DISCUSSION THREAD

**Discussion by LaZsolt** [dbuezas/lgt8fx#242] [Show and tell]
LTG8F328x has 7 more digital I/O-s (port E) **in LQFP32 package** than ATmega328P. Most of this I/O-s not exposed on any nano style or mini style boards.

Ports E0, E2 or E6 has primary alternate function. When you set these ports with `pinMode()` the primary functions will be disabled.

![kép](https://user-images.githubusercontent.com/60486174/214576930-194733d2-b2b1-47c6-b834-ce4bd6b59ba9.png)

But if someone want to use port C6 (Pin 29) the RESET function must be disabled before set the port with `pinMode()` because it does not do that. After that the hardware RESET will not work and cannot be reprogrammed through serial port (Nor USB-Serial.) In this case only the power on RESET will work.

Example code:
```
uint8_t resdis = PMX2 | _BV(C6EN);
PMX2 = 0x80;
PMX2 = resdis;
pinMode(C6,OUTPUT);      // equivalent pinMode(D27,OUTPUT);

```

**Comment by dbuezas** [comment]
Do you know if something similar is required before using any of `e4`,`c0`,`f1`,`f2`,`f3`,`f5`,`f6`,`f7`?
I may need to add that to the [Arduino Web Timers App](https://dbuezas.github.io/arduino-web-timers)

**Reply by LaZsolt** [reply]
The ports you mentioned have no default alternate function which should be disabled before use.
On the other hand, I was wrong first, `pinMode()` works better than I what wrote above. So I made a correction.

**Comment by LaZsolt** [comment]
When I started this discussion, I was wrong.
pinMode() works as it should. So I corrected the text.

**Reply by dbuezas** [reply]
Oh gotcha.
May be worth mentioning that if they share a pin with another gpio, the other one must be kept as input (or fully deactivated when possible). Otherwise the internal short will break the chip very quickly.

**Reply by LaZsolt** [reply]
When LQFP32 package selected, `pinMode()` can not enable drive ports shared pins, but timers outputs may interfere port D1, D2, D6 or E5.
There is a greater chance of a short circuit between internally connected ports when using SSOP20 package.

**Reply by jayzakk** [reply]
Hey guys,

I remember modifying pinmode after adding the E ports and a thought of adding C6 also. I decided to not auto-change C6 alt mode out of fear users do make mistake or don't really understand the consequences and may end up needing a programmer to use their MCU again - IF they kept SWD working. It's kind of a dangerous switch, and having all alternate modes enabled, I am not really sure if there is a way to re-program the LGTF ever again. A long (long!) forced delay before C6 mode may migitate that issue.

It's a one-way function only, as there doesn't exist any function to re-enable the default again. This rather won't be needed: if someone choose to use this pins as GPIO, the hardware design choice forbids the use of the original modes anyways.

Regarding the LQPF32 shared pins, i remember having seen something in the datasheet about the mode automatically changing when setting the appropriate bits in the timer modules. But I don't have this package to check it and if there really is need of changing other registers.

A function to en/disabling 80mA pin mode (maybe with keeping track of the max# of enabled ones to help prevent reaching max package power limit?) and the alternate pins for Rx/Tx could be helpful.

Together with the new read-vccm, it may be a good idea to add a page to the wiki to address documentation to this special LGTF functions() and limits available to users. Virtually anyone seem to be confused of the two-in-one pin text of that small 32 boards, also. 

Nobody should be forced to read through all of that datasheet 😂
Guess we may find even more where some simple function could be added. 

Writing this, I still wonder about the datasheet announced feature of code protection when there are no fuses to set, but this may be some other topic or just copy/paste errata 🤷

Some more docs or inlines/defines for the coprocessor could be helpful, too - or a modified version of the compiler to make them available for everyone? Lots of possible work to do....

(Yeah, I'm still alive, but had some hard and busy times at job... Thanks a lot for the 1.0.7!)

**Reply by pixelwaster** [reply]
Is there anyone interested in the inlines/defines for the coprocessor besides [jayzakk]?  Sorry for the high jack, but this is only the second time I've seen this mentioned about LGT chip.

**Reply by dbuezas** [reply]
Oh, yes that would be super interesting to have. I never got that to work
